subprojects/docs/src/docs/userguide/more_about_tasks.adoc

```diff
@@ -831,112 +947,175 @@ Gradle also considers the _code_ of the task as part of the inputs to the task.
+Note that if a task has an output directory specified, any files added to that directory since the last time it was executed are ignored and will NOT cause the task to be out of date. This is so unrelated tasks may share an output directory without interfering with each other. If this is not the behaviour you want for some reason, consider using link:{javadocPath}/org/gradle/api/tasks/TaskOutputs.html#upToDateWhen-groovy.lang.Closure-[TaskOutputs.upToDateWhen(groovy.lang.Closure)]
 
-The inputs for the task are also used to calculate the <<build_cache,build cache>> key used to load task outputs when enabled. For more details see <<sec:task_output_caching>>.
+The inputs for the task are also used to calculate the <<build_cache.adoc#build_cache,build cache>> key used to load task outputs when enabled. For more details see <<build_cache.adoc#sec:task_output_caching,Task output caching>>.
 
 [[sec:advanced_inc_build]]
-==== Advanced techniques
+=== Advanced techniques
 
 Everything you’ve seen so far in this section will cover most of the use cases you’ll encounter, but there are some scenarios that need special treatment. We’ll present a few of those next with the appropriate solutions.
 
 
 [[sec:add_cached_input_output_methods]]
-===== Adding your own cached input/output methods
+==== Adding your own cached input/output methods
 
 Have you ever wondered how the `from()` method of the `Copy` task works? It’s not annotated with `@InputFiles` and yet any files passed to it are treated as formal inputs of the task. What’s happening?
 
 The implementation is quite simple and you can use the same technique for your own tasks to improve their APIs. Write your methods so that they add files directly to the appropriate annotated property. As an example, here’s how to add a `sources()` method to the custom `ProcessTemplates` class we introduced earlier:
 
-++++
-<sample id="incrementalBuildCustomMethods" dir="userguide/tasks/incrementalBuild/incrementalBuildAdvanced" title="Declaring a method to add task inputs">
-                    <sourcefile file="buildSrc/src/main/java/org/example/ProcessTemplates.java" snippet="custom-task-class"/>
-                    <sourcefile file="build.gradle" snippet="custom-task-class"/>
-                    <output args="processTemplates" ignoreExtraLines="true" ignoreLineOrder="true"/>
-                </sample>
-++++
+===== Example: Declaring a method to add task inputs
+
+.ProcessTemplates.java
+[source,java]
+----
+include::{samplesPath}/userguide/tasks/incrementalBuild/incrementalBuildAdvanced/buildSrc/src/main/java/org/example/ProcessTemplates.java[tag=task-arg-method]
```

Comment:
The source included here looks really different from the old doc. Seems like we're using the wrong tag?

Reply:
Yep. Fixed